src/main/kotlin/org/rust/ide/inspections/RsAssignToImmutableInspection.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemsHolder
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.AssignmentOp
+import org.rust.lang.core.psi.ext.operatorType
+import org.rust.lang.core.types.isDereference
+import org.rust.lang.core.types.isMutable
+import org.rust.lang.core.types.ty.TyPointer
+import org.rust.lang.core.types.ty.TyReference
+import org.rust.lang.core.types.type
+import org.rust.lang.core.types.unwrapParenExprs
+
+class RsAssignToImmutableInspection : RsLocalInspectionTool() {
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitBinaryExpr(expr: RsBinaryExpr) {
+                if (expr.isAssignBinaryExpr()) checkAssignment(expr, holder)
+            }
+        }
+
+    private fun checkAssignment(expr: RsBinaryExpr, holder: ProblemsHolder) {
+        val left = unwrapParenExprs(expr.left)
+        if (left.isMutable) return
+
+        when (left) {
+            is RsDotExpr -> registerProblem(holder, expr, "field of immutable binding")
+            is RsIndexExpr -> registerProblem(holder, expr, "indexed content of immutable binding")
+            is RsUnaryExpr ->
+                if (left.isDereference) when (left.expr?.type) {
+                    is TyReference -> registerProblem(holder, expr, "immutable borrowed content")
+                    is TyPointer -> registerProblem(holder, expr, "immutable dereference of raw pointer")
+                }
+        }
+    }
+
+    private fun registerProblem(holder: ProblemsHolder, expr: RsExpr, message: String) {
+        holder.registerProblem(expr, "Cannot assign to $message [E0594]")
+    }
+}
+
+private fun RsExpr?.isAssignBinaryExpr(): Boolean {
+    val op = this as? RsBinaryExpr ?: return false
```

Comment:
Maybe `if (this !is RsBinaryExpr) return false`? Don't know what is more readable though

Reply:
Maybe it is better to return `this is BinaryExpr && this.operatorType is AssignmentOp`?